File: src/org/jetbrains/agda/scope/AgdaGlobalScope.java
package org.jetbrains.agda.scope;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import org.jetbrains.agda.mixfix.Grammar;
import org.jetbrains.agda.mixfix.TreeElement;
import org.jetbrains.agda.psi.*;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

public class AgdaGlobalScope {
    public static PsiElement findDeclaration(@NotNull AgdaReferenceElementImpl element) {
        if (element.getParent() instanceof Application) {
            TreeElement term = Grammar.parse((Application) element.getParent());
            if (term != null) {
                PsiElement declaration = term.getDeclaration(element);
                if (declaration != null) {
                    return declaration;
                }
            }
        }
        Map<String, PsiElement> declarations = new AgdaExpressionScope(element).getVisibleDeclarations();
        String text = ((AName) element).getId().getText();
        return declarations.get(text);
    }

    private static void getDeclarations(PsiElement element, Map<String, PsiElement> map) {
        if (element instanceof DataDeclaration) {
            DataDeclaration dataDeclaration = ((DataDeclaration) element);
            map.put(dataDeclaration.getNameDeclaration().getText(), element);
        }
        if (element instanceof Constructors) {
            Constructors constructors = ((Constructors) element);
            for (TypeSignature constructor: constructors.getTypeSignatureList()) {
                for (NameDeclaration nameDeclaration: constructor.getNameDeclarationList()) {
                    map.put(nameDeclaration.getText(), nameDeclaration);
                }
            }
        }
        if (element instanceof FunctionTypeDeclaration) {
            String text = ((FunctionTypeDeclaration) element).getFirstChild().getText();
            map.put(text, element);
        }
        for (PsiElement child : element.getChildren()) {
            getDeclarations(child, map);
        }
    }

    public static Map<String, PsiElement> getDeclarations(PsiElement element) {
        PsiElement root = element;
        while (!(root instanceof PsiFile)) {
            root = root.getParent();
        }
        Map<String, PsiElement> map = new HashMap<String, PsiElement>();
        getDeclarations(root, map);
        return map;
    }
}

File: src/org/jetbrains/agda/scope/AgdaGlobalScope.kt
package org.jetbrains.agda.scope

import org.jetbrains.agda.psi.FqName
import com.intellij.psi.PsiElement
import java.util.Arrays
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiFile

/**
 * @author Evgeny.Kurbatsky
 */
public class AgdaGlobalScope() {

        public fun find(element : FqName) : PsiElement? {
            val text : String = element.getText()!!
            val path : Array<String> = text.split("\\.")
            var containingDirectory : PsiDirectory? = element.getContainingFile()?.getContainingDirectory();
            while (containingDirectory!!.getName() != "src") {
                containingDirectory = containingDirectory?.getParent();
            }
            val file : PsiFile? = findPath(containingDirectory, path, 0)
            return file
        }

        private fun findPath(directory : PsiDirectory?, path : Array<String>, index : Int) : PsiFile? {
            var name : String? = path.get(index)
            if (path.size == index + 1)
            {
                for (file : PsiFile? in directory?.getFiles()!!)
                {
                    var fileName : String? = file?.getName()
                    if (fileName?.equals(name + ".agda")!!)
                    {
                        return file
                    }

                }
            }
            else
            {
                for (psiDirectory : PsiDirectory? in directory?.getSubdirectories()!!)
                {
                    var fileName : String? = psiDirectory?.getName()
                    if (fileName?.equals(name)!!)
                    {
                        return findPath(psiDirectory, path, index + 1)
                    }

                }
            }
            return null
        }

}
